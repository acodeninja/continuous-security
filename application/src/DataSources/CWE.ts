import {DOMParser} from '@xmldom/xmldom';
import XmlReader from 'xml-reader';
import XmlQuery from 'xml-query';
import {select} from 'xpath';
import {tidyString} from '../Helpers';

import CWEDataset from '../assets/cwe.xml';

export class CWEDoesNotExist extends Error {
  message = 'Could not find CWE';
}

export class CWE {
  private data = (new DOMParser({
    errorHandler: undefined,
  })).parseFromString(CWEDataset);

  getById(id: string): ReportOutputIssueReference {
    const parsedId = id.toLowerCase().replace('cwe-', '');

    const [weakness] = select(`//*[local-name(.)='Weakness' and @id='${parsedId}']`, this.data);

    if (weakness) {
      const queryable = XmlQuery(XmlReader.parseSync(weakness.toString()));
      const attributes = queryable.attr() as {
        ID: string;
        Name: string;
        Abstraction: string;
        Structure: string;
        Status: string;
        xmlns: string;
      };

      return {
        label: id,
        title: attributes.Name,
        description: this.expandCWELinks(tidyString(queryable.find('Description').first().text())),
        directLink: `https://cwe.mitre.org/data/definitions/${parsedId}.html`,
        dataSourceSpecific: {
          cwe: {
            extendedDescription: this.expandCWELinks(tidyString(queryable.find('Extended_Description').first().text())),
            background: this.expandCWELinks(tidyString(queryable.find('Background_Detail').first().text())),
            consequences: this.expandConsequences(queryable.find('Consequence')),
            mitigations: this.expandMitigations(queryable.find('Mitigation')),
          },
        },
      };
    }

    throw new CWEDoesNotExist();
  }

  private expandConsequences(consequencesNodes): ReportOutputIssueReference['dataSourceSpecific']['cwe']['consequences'] {
    const consequences = [];

    consequencesNodes.each(node => {
      node = XmlQuery(node);
      const consequence = {};
      const scopes = node.find('Scope').children().map(n => tidyString(n.value));
      const impacts = node.find('Impact').children().map(n => tidyString(n.value));
      const likelihood = tidyString(node.find('Likelihood').text());
      const note = tidyString(node.find('Note').text());

      consequence['scopeImpacts'] = scopes.map((scope, index) => {
        const scopeImpact = {scope};
        if (impacts[index]) scopeImpact['impact'] = impacts[index];
        return scopeImpact;
      });

      if (likelihood) consequence['likelihood'] = likelihood;
      if (note) consequence['note'] = note;

      consequences.push(consequence);
    });

    return consequences;
  }

  private expandMitigations(mitigationsNodes) {
    const mitigations = [];

    mitigationsNodes.each(node => {
      node = XmlQuery(node);

      mitigations.push({
        phase: node.find('Phase').children().map(p => tidyString(p.value)).join(' / '),
        description: tidyString(node.find('Description').text()),
        effectiveness: tidyString(node.find('Effectiveness').text()),
        notes: tidyString(node.find('Effectiveness_Notes').text()),
      });
    });

    return mitigations;
  }

  private expandCWELinks(text: string): string {
    return text.replace(/(CWE-)(\d+)/gm, '[$1$2](https://cwe.mitre.org/data/definitions/$2.html)');
  }
}
