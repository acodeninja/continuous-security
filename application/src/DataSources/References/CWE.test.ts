import {CWE, CWEDoesNotExist} from './CWE';

describe('with a CWE id that exists', () => {
  const reference= (new CWE).getById('CWE-1004');

  test('has the correct label', () => {
    expect(reference.label).toEqual('CWE-1004');
  });

  test('has the correct title', () => {
    expect(reference.title).toEqual('Sensitive Cookie Without \'HttpOnly\' Flag');
  });

  test('has the correct description', () => {
    expect(reference.description).toEqual('The product uses a cookie to store ' +
      'sensitive information, but the cookie is not marked with the HttpOnly flag.');
  });

  test('has the correct url', () => {
    expect(reference.directLink).toEqual('https://cwe.mitre.org/data/definitions/1004.html');
  });

  test('includes the cwe specific data', () => {
    expect(reference.dataSourceSpecific).toHaveProperty('cwe');
  });

  test('has the correct extended description', () => {
    expect(reference.dataSourceSpecific.cwe.extendedDescription).toMatchSnapshot();
  });

  test('has the correct background', () => {
    expect(reference.dataSourceSpecific.cwe.background).toMatchSnapshot();
  });

  test('has the correct list of consequences', () => {
    expect(reference.dataSourceSpecific.cwe.consequences).toEqual(expect.arrayContaining([{
      scopeImpacts: [{scope: 'Confidentiality', impact: 'Read Application Data'}],
      note: 'If the HttpOnly flag is not set, then sensitive information stored in the cookie ' +
        'may be exposed to unintended parties.',
    }]));
    expect(reference.dataSourceSpecific.cwe.consequences).toEqual(expect.arrayContaining([{
      scopeImpacts: [{scope: 'Integrity', impact: 'Gain Privileges or Assume Identity'}],
      note: 'If the cookie in question is an authentication cookie, then not setting the ' +
        'HttpOnly flag may allow an adversary to steal authentication data (e.g., a session ID) ' +
        'and assume the identity of the user.',
    }]));
  });

  test('has the correct list of mitigations', () => {
    expect(reference.dataSourceSpecific.cwe.mitigations).toEqual(expect.arrayContaining([{
      phase: 'Implementation',
      description: 'Leverage the HttpOnly flag when setting a sensitive cookie in a response.',
      effectiveness: 'High',
      notes: 'While this mitigation is effective for protecting cookies from a browser\'s own ' +
        'scripting engine, third-party components or plugins may have their own engines that ' +
        'allow access to cookies. Attackers might also be able to use XMLHTTPResponse to read ' +
        'the headers directly and obtain the cookie.',
    }]));
  });
});

describe('with a CWE id that does not exist', () => {
  test('throws a CWEDoesNotExist error', () => {
    expect(() => (new CWE).getById('00000')).toThrow(CWEDoesNotExist);
  });
});
