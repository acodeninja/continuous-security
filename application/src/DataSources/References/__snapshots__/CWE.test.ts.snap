// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with a CWE id that exists has the correct background 1`] = `"An HTTP cookie is a small piece of data attributed to a specific website and stored on the user's computer by the user's web browser. This data can be leveraged for a variety of purposes including saving information entered into form fields, recording user activity, and for authentication purposes. Cookies used to save or record information generated by the user are accessed and modified by script code embedded in a web page. While cookies used for authentication are created by the website's server and sent to the user to be attached to future requests. These authentication cookies are often not meant to be accessed by the web page sent to the user, and are instead just supposed to be attached to future requests to verify authentication details."`;

exports[`with a CWE id that exists has the correct extended description 1`] = `"The HttpOnly flag directs compatible browsers to prevent client-side script from accessing cookies. Including the HttpOnly flag in the Set-Cookie HTTP response header helps mitigate the risk associated with Cross-Site Scripting (XSS) where an attacker's script code might attempt to read the contents of a cookie and exfiltrate information obtained. When set, browsers that support the flag will not reveal the contents of the cookie to a third party via client-side script executed via XSS."`;
